incubator/virtualcluster/pkg/resource-syncer/controller/controller.go

```diff
@@ -0,0 +1,212 @@
+	if options.reconciler == nil {
+		return nil, fmt.Errorf("must specify Reconciler")
+	}
+
+	if len(name) == 0 {
+		return nil, fmt.Errorf("must specify Name for Controller")
+	}
+
+	c := &MultiClusterController{
+		name:       name,
+		clusters:   make(map[Cluster]struct{}),
+		Options:    options,
+	}
+
+	if c.JitterPeriod == 0 {
+		c.JitterPeriod = 1 * time.Second
+	}
+
+	if c.MaxConcurrentReconciles <= 0 {
+		c.MaxConcurrentReconciles = 1
+	}
+
+	if c.Queue == nil {
+		c.Queue = workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter())
+	}
+
+	return c, nil
+}
+
+// WatchOptions is used as an argument of WatchResource methods (just a placeholder for now).
+// TODO: consider implementing predicates.
+type WatchOptions struct {
+}
+
+// WatchClusterResource configures the Controller to watch resources of the same Kind as objectType,
+// in the specified cluster, generating reconcile Requests from the Cluster's context
+// and the watched objects' namespaces and names.
+func (c *MultiClusterController) WatchClusterResource(cluster Cluster, o WatchOptions) error {
+	c.clusters[cluster] = struct{}{}
+	h := &handler.EnqueueRequestForObject{Cluster: cluster.GetClientInfo(), Queue: c.Queue}
+	return cluster.AddEventHandler(c.objectType, h)
+}
+
+// Start starts the ClustersController's control loops (as many as MaxConcurrentReconciles) in separate channels
+// and blocks until an empty struct is sent to the stop channel.
+func (c *MultiClusterController) Start(stop <-chan struct{}) error {
+	// pre start all the cluster caches
+	wg := &sync.WaitGroup{}
+	wg.Add(len(c.clusters))
+
+	errCh := make(chan error)
+	for cl := range c.clusters {
+		go func(cl Cluster) {
+			if err := cl.Start(stop); err != nil {
+				errCh <- err
+			}
+		}(cl)
+
+		go func(cl Cluster) {
+			defer wg.Done()
+
+			if ok := cl.WaitForCacheSync(stop); !ok {
+				errCh <- fmt.Errorf("failed to wait for caches to sync")
+			}
+		}(cl)
+	}
+
+	wg.Wait()
+
+	klog.Infof("start clusters-controller %q", c.name)
+
+	defer c.Queue.ShutDown()
+
+	for i := 0; i < c.MaxConcurrentReconciles; i++ {
+		go wait.Until(func() {
+			for c.processNextWorkItem() {
+			}
+		}, c.JitterPeriod, stop)
+	}
+
+	select {
+	case <-stop:
+		return nil
+	case err := <-errCh:
+		return err
+	}
+}
+
+func (c *MultiClusterController) processNextWorkItem() bool {
+	obj, shutdown := c.Queue.Get()
+	if obj == nil {
+		c.Queue.Forget(obj)
+	}
+
+	if shutdown {
+		klog.Warning("Shutting down. Ignore work item and stop working.")
+		return false
+	}
+
+	defer c.Queue.Done(obj)
+	var req reconciler.Request
+	var ok bool
+	if req, ok = obj.(reconciler.Request); !ok {
+		klog.Warning("Work item is not a Request. Ignore it. Next.")
+		c.Queue.Forget(obj)
+		return true
+	}
+
+	if result, err := c.reconciler.Reconcile(req); err != nil {
+		klog.Error(err)
+		klog.Error("Could not reconcile Request. Stop working.")
+		c.Queue.AddRateLimited(req)
+		return false
```

Comment:
Do you permanently lose one reconciler thread if Reconcile func returns false?

Reply:
`go wait.Until(func() {
    for c.processNextWorkItem() {
    }
}, c.JitterPeriod, stop)`
IIUC this just means the for loop will break, the reconciler thread will not return, wait.Until will call worker func again after JitterPeriod.

Reply:
Got you